### PR TITLE
TimePathedSource helper methods

### PR DESCRIPTION
**Pull request by rubanm**
Moving some re-usable logic to the companion object.

### CODE REVIEW

**Comment by isnotinvain** [comment]
+1

File: scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala [hunk 1]
```diff
@@ -31,6 +31,38 @@ object TimePathedSource {
       "%1$tm" -> Months(1)(tz), "%1$tY" -> Years(1)(tz))
       .find { unitDur: (String, Duration) => pattern.contains(unitDur._1) }
       .map(_._2)
+
+  /**
+   * Gives all paths in the given daterange with windows based on the provided duration.
+   */
+  def allPathsWithDuration(pattern: String, duration: Option[Duration], dateRange: DateRange, tz: TimeZone): Iterable[String] =
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala]: I don't like passing an Option just to call .getOrElse(Nil). I'd rather require the duration here and have callers do that.

File: scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala [hunk 2]
```diff
@@ -31,6 +31,38 @@ object TimePathedSource {
...
+          this.toPath(pattern, dr.start, tz)
+        }
+    }.getOrElse(Nil)
+
+  /**
+   * Gives all read paths in the given daterange.
+   */
+  def readPathsFor(pattern: String, dateRange: DateRange, tz: TimeZone): Iterable[String] = {
+    val stepSize = TimePathedSource.stepSize(pattern, DateOps.UTC)
+    this.allPathsWithDuration(pattern, stepSize, dateRange, DateOps.UTC)
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala]: `this` is weird on an object.

File: scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala [hunk 3]
```diff
@@ -31,6 +31,38 @@ object TimePathedSource {
...
+  }
+
+  /**
+   * Gives the write path based on daterange end.
+   */
+  def writePathFor(pattern: String, dateRange: DateRange, tz: TimeZone): String = {
+    // TODO this should be required everywhere but works on read without it
+    // maybe in 0.9.0 be more strict
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala]: we are well past 0.9.0 now. Need to update this comment.

File: scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala [hunk 4]
```diff
@@ -31,6 +31,38 @@ object TimePathedSource {
...
+
+  /**
+   * Gives all paths in the given daterange with windows based on the provided duration.
+   */
+  def allPathsWithDuration(pattern: String, duration: Option[Duration], dateRange: DateRange, tz: TimeZone): Iterable[String] =
+    duration.map { dur =>
+      // This method is exhaustive, but too expensive for Cascading's JobConf writing.
+      dateRange.each(dur)
+        .map { dr: DateRange =>
+          this.toPath(pattern, dr.start, tz)
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala]: `this` is not needed here, right?

File: scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala (hunk 3 above)
Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala]: Any idea why this was required?

Comment (rubanm) [inline scalding-core/src/main/scala/com/twitter/scalding/TimePathedSource.scala]: Seems to me like this is actually needed to globify read paths but has to be removed for the write path?

**Comment by rubanm** [comment]
@johnynek addressed your comments. Let me know if we should clarify anything for the "/*" logic. Thanks.

**Pull request merged**